File: 2017/09/stream-processing-2.go
package main

import "fmt"
import "os"
import "io/ioutil"
import "strings"


type StateError struct {
    char rune
    transition string
}

func (e *StateError) Error() string {
    return fmt.Sprintf("bad char '%v' for transition %s", e.char, e.transition)
}


type Transition func(state *state) (Transition, error)


type state struct {
    Sizes []int
    Garbage int
    Stream chan rune
}


func NewState(data string) *state {
    var stream = make(chan rune)
    go func() {
        defer close(stream)
        for _, char := range data {
            stream <- char
        }
        stream <- 0
    }()
    return &state{Stream: stream}
}


func readInputFile(filename string) string {
    data, err := ioutil.ReadFile(filename)
    if err != nil {
        panic(err)
    }

    return strings.Trim(string(data), "\n")
}


func start(state *state) (Transition, error) {
    char := <-state.Stream
    switch {
    case char == '{':
        return newGroup, nil
    default:
    }
    return nil, &StateError{char, "start"}
}


func newGroup(state *state) (Transition, error) {
    char := <-state.Stream
    switch {
    case char == '{':
        return newGroup, nil
    case char == '<':
        return garbage, nil
    case char == '}':
        return endGroup, nil
    default:
    }
    return nil, &StateError{char, "newGroup"}
}


func garbage(state *state) (Transition, error) {
    char := <- state.Stream
    switch {
    case char == '!':
        return ignore, nil
    case char == '>':
        return endGarbage, nil
    case char == 0:
        return nil, &StateError{char, "garbage"}
    default:
        state.Garbage += 1
        return garbage, nil
    }
}


func ignore(state *state) (Transition, error) {
    _ = <- state.Stream
    return garbage, nil
}


func endGarbage(state *state) (Transition, error) {
    state.Sizes = append(state.Sizes, state.Garbage)
    state.Garbage = 0
    char := <- state.Stream
    switch {
    case char == ',':
        return nextThing, nil
    case char == '}':
        return endGroup, nil
    default:
        return nil, &StateError{char, "endGarbage"}
    }
}


func nextThing(state *state) (Transition, error) {
    char := <- state.Stream
    switch {
    case char == '{':
        return newGroup, nil
    case char == '<':
        return garbage, nil
    default:
        return nil, &StateError{char, "nextThing"}
    }
}


func endGroup(state *state) (Transition, error) {
    char := <-state.Stream
    switch {
    case char == 0:
        return nil, nil  // we're done
    case char == '{':
        return newGroup, nil
    case char == ',':
        return nextThing, nil
    case char == '}':
        return endGroup, nil
    default:
        return nil, &StateError{char, "endGroup"}
    }
}


func getSizes(stream string) []int {
    var state = NewState(stream)

    transition, err := start(state)
    for transition != nil || err != nil {
        transition, err = transition(state)
    }
    if err != nil {
        panic(err)
    }
    return state.Sizes
}


func sum(numbers []int) (total int) {
    for _, n := range numbers {
        total += n
    }
    return
}


func main() {
    var filename = os.Args[1]
    var stream = readInputFile(filename)
    var sizes = getSizes(stream)
    fmt.Println(sum(sizes))
}

File: 2017/09/stream-processing-2.py
# -*- coding: utf-8 -*-
import sys


class StateError(Exception):
    def __init__(self, transition, char):
        super().__init__(f'found {repr(char)} in {transition}')


def read_input_file(filename):
    with open(filename) as f:
        content = f.read().strip()
    return iter(content)


class State:
    def __init__(self, stream):
        self.stream = iter(stream)
        self.scores = []
        self.garbage = 0


def count_garbage_size(stream):
    state = State(stream)
    transition = start(state)
    while transition:
        transition = transition(state)
    return state.scores


def start(state):
    char = next(state.stream)
    if char == '{':
        return new_group
    raise StateError('start', char)


def new_group(state):
    char = next(state.stream)
    if char == '{':
        return new_group
    elif char == '<':
        return garbage
    elif char == '}':
        return end_group
    raise StateError('group', char)


def garbage(state):
    char = next(state.stream)
    if char == '!':
        return ignore
    elif char == '>':
        state.scores.append(state.garbage)
        state.garbage = 0
        return end_garbage
    state.garbage += 1
    return garbage


def ignore(state):
    next(state.stream)
    return garbage


def end_garbage(state):
    char = next(state.stream)
    if char == ',':
        return next_group
    elif char == '}':
        return end_group
    raise StateError('end_garbage', char)


def next_group(state):
    char = next(state.stream)
    if char == '{':
        return new_group
    elif char == '<':
        return garbage
    raise StateError('next_group', char)


def end_group(state):
    try:
        char = next(state.stream)
    except StopIteration:
        return None

    if char == '{':
        return new_group
    elif char == ',':
        return next_group
    elif char == '}':
        return end_group

    raise StateError('end_group', char)


def main(filename):
    stream = read_input_file(filename)
    sizes = count_garbage_size(stream)
    print(sum(sizes))


if __name__ == '__main__':
    main(sys.argv[1])
